waspc/test/Analyzer/TypeChecker/InternalTest.hs

```diff
@@ -0,0 +1,202 @@
+module Analyzer.TypeChecker.InternalTest where
+
+import qualified Analyzer.Parser as P
+import Analyzer.Type
+import Analyzer.TypeChecker.AST
+import Analyzer.TypeChecker.Internal
+import Analyzer.TypeChecker.Monad (Bindings, runT, runTWithBound)
+import Analyzer.TypeChecker.TypeError
+import qualified Analyzer.TypeDefinitions as TD
+import qualified Data.HashMap.Strict as H
+import Data.List.NonEmpty (NonEmpty ((:|)))
+import Test.Tasty.Hspec
+import Test.Tasty.QuickCheck
+
+-- TODO:
+-- Having Arbitrary instances for Types and TypedAST would allow much more thorough testing
+-- of the type checker, and is probably worth the effort to figure out how to make instances
+-- given how complex the code is, how many corner cases there are, and how well-defined the
+-- required properties of the type-checking functions are.
+
+chooseType :: Gen Type
+chooseType =
+  elements
+    [ StringType,
+      NumberType,
+      BoolType,
+      ExtImportType,
+      QuoterType "json",
+      QuoterType "psl"
+    ]
+
+checkExpr' :: Bindings -> P.Expr -> Either TypeError TypedExpr
+checkExpr' bindings expr = runTWithBound bindings TD.empty $ checkExpr expr
+
+spec_Parser :: Spec
+spec_Parser = do
+  describe "Analyzer.TypeChecker.Internal" $ do
+    describe "unify" $ do
+      it "Doesn't affect 2 expressions of the same type" $ do
+        property $ \(a, b) ->
+          let initial = IntegerLiteral a :| [DoubleLiteral b]
+              actual = unify initial
+           in actual == Right (initial, NumberType)
+      it "Unifies two same-typed dictionaries to their original type" $ do
+        let typ = DictType $ H.fromList [("a", DictRequired BoolType), ("b", DictOptional NumberType)]
+        let a = Dict [("a", BoolLiteral True), ("b", IntegerLiteral 2)] typ
+        let b = Dict [("a", BoolLiteral True), ("b", DoubleLiteral 3.14)] typ
+        unify (a :| [b])
+          `shouldBe` Right (a :| [b], typ)
+      it "Unifies an empty dict and a dict with one property" $ do
+        let a = Dict [] (DictType H.empty)
+        let b = Dict [("a", BoolLiteral True)] $ DictType $ H.singleton "a" $ DictRequired BoolType
+        let expected = DictType $ H.singleton "a" $ DictOptional BoolType
+        fmap (fmap exprType . fst) (unify (a :| [b]))
+          `shouldBe` Right (expected :| [expected])
+      it "Is idempotent when unifying an empty dict and a singleton dict" $ do
+        let a = Dict [] (DictType H.empty)
+        let b = Dict [("a", BoolLiteral True)] $ DictType $ H.singleton "a" $ DictRequired BoolType
+        unify (a :| [b]) `shouldBe` (unify (a :| [b]) >>= unify . fst)
+
+    describe "checkExpr" $ do
+      it "Types string literals as StringType" $ do
```

Comment:
Lot of repetition, could we remove it by doing something similar as we did in previous PR -> writing a `test` function that returns `it ...` and then calling it multiple times but with different arguments?